### PR TITLE
Test case for recovery from disk deletion over vSphere

### PR DESCRIPTION
**Pull request by ebenahar**
- Automation for disk replacement procedure detailed in https://bugzilla.redhat.com/show_bug.cgi?id=1787236#c16
- Addresses issues:
  - https://github.com/red-hat-storage/ocs-ci/issues/1080
  - https://github.com/red-hat-storage/ocs-ci/issues/1814

Signed-off-by: Elad Ben Aharon <ebenahar@redhat.com>

### CODE REVIEW

File: ocs_ci/ocs/platform_nodes.py [hunk 1]
```diff
@@ -73,12 +73,12 @@ def restart_nodes(self, nodes, force=True):
             "Restart nodes functionality is not implemented"
         )
 
-    def detach_volume(self, node):
+    def detach_volume(self, volume, node):
```
Comment (RazTamir) [inline ocs_ci/ocs/platform_nodes.py:76]: Applies to all function's signature changes below, please check that no usage will cause backward compatibility failures

File: ocs_ci/ocs/resources/pod.py [hunk 2]
```diff
@@ -714,6 +714,22 @@ def get_pods_having_label(label, namespace):
...
+def get_deployments_having_label(label, namespace):
+    """
+    Fetches deployment resources with given label in given namespace
+
+    Args:
+        label (str): label which deployments might have
+        namespace (str): Namespace in which to be looked up
+
+    Return:
+        dict: of deployment info
```
Comment (RazTamir) [inline ocs_ci/ocs/resources/pod.py:726]: list?
'items' is a list of dicts iirc

File: ocs_ci/ocs/resources/pod.py [hunk 3]
```diff
@@ -790,6 +806,25 @@ def get_osd_pods(osd_label=constants.OSD_APP_LABEL, namespace=None):
...
+            (default: defaults.OSD_APP_LABEL)
+        namespace (str): Namespace in which ceph cluster lives
+            (default: defaults.ROOK_CLUSTER_NAMESPACE)
+
+    Returns:
+        list : of osd deployment OCS objects
+    """
+    namespace = namespace or config.ENV_DATA['cluster_namespace']
+    osds = get_deployments_having_label(osd_label, namespace)
+    osd_deployments = [OCS(**osd) for osd in osds]
```
Comment (RazTamir) [inline ocs_ci/ocs/resources/pod.py:824]: confirms previous comment

File: ocs_ci/utility/vsphere.py [hunk 4]
```diff
@@ -587,3 +587,57 @@ def destroy_folder(self, name, cluster, dc):
...
+
+        Args:
+            volume_path (str): Volume path in the datastore (i.e,
+                '[vsanDatastore] d4210a5e-40ce-efb8-c87e-040973d176e1/control-plane-1.vmdk')
+            vm (vm): VM (vm) object
+
+        """
+        _, volume = self.get_vm_and_volume_by_volume_path([vm], volume_path)
+
+        spec = vim.vm.ConfigSpec()
```
Comment (RazTamir) [inline ocs_ci/utility/vsphere.py:632]: vim?

File: ocs_ci/utility/vsphere.py [hunk 5]
```diff
@@ -587,3 +587,57 @@ def destroy_folder(self, name, cluster, dc):
...
+        device_changes = []
+        disk_spec = vim.vm.device.VirtualDeviceSpec()
```
Comment (RazTamir) [inline ocs_ci/utility/vsphere.py:634]: ok 'vim' is not a type, so please use upper case when calling an object

File: tests/manage/z_cluster/nodes/test_disk_failures.py [hunk 6]
```diff
@@ -129,3 +134,104 @@ def test_detach_attach_2_data_volumes(self, nodes, pvc_factory, pod_factory):
         # Validate cluster is still functional
         self.sanity_helpers.health_check()
         self.sanity_helpers.create_resources(pvc_factory, pod_factory)
+
+    @vsphere_platform_required
+    @pytest.mark.polarion_id("OCS-0000")
```
Comment (RazTamir) [inline tests/manage/z_cluster/nodes/test_disk_failures.py:139]: don't forget to create a Polarion TC

File: tests/manage/z_cluster/nodes/test_disk_failures.py [hunk 7]
```diff
@@ -129,3 +134,104 @@ def test_detach_attach_2_data_volumes(self, nodes, pvc_factory, pod_factory):
...
+        # Get the node that has the OSD pod running on
+        logger.info(f"Getting the node that has the OSD pod {osd_pod.ocp.resource_name} running on")
+        osd_node = get_pod_node(osd_pod)
+        volume_size = osd_pvc.size
+
+        # Get the corresponding OSD deployment
+        logger.info(f"Getting the corresponding OSD deployment for OSD PVC {claim_name}")
+        osd_deployment = [
+            osd_pod for osd_pod in get_osd_deployments() if osd_pod.get()
+            .get('metadata').get('labels').get('ceph.rook.io/pvc') == claim_name
```
Comment (RazTamir) [inline tests/manage/z_cluster/nodes/test_disk_failures.py:183]: 'ceph.rook.io/pvc' --> constant

File: ocs_ci/ocs/platform_nodes.py (hunk 1 above)
Comment (ebenahar) [reply ocs_ci/ocs/platform_nodes.py:76]: ack. Will change

File: ocs_ci/ocs/resources/pod.py (hunk 2 above)
Comment (ebenahar) [reply ocs_ci/ocs/resources/pod.py:726]: will fix

File: ocs_ci/ocs/resources/pod.py (hunk 3 above)
Comment (ebenahar) [reply ocs_ci/ocs/resources/pod.py:824]: ack

File: ocs_ci/utility/vsphere.py (hunk 4 above)
Comment (ebenahar) [reply ocs_ci/utility/vsphere.py:632]: 'vim' is imported from pyVmomi module. See line 9 in this file
In any case, this function will be deleted in upcoming commit as I have incorporated its implementation into the remove_disk function in this file

File: ocs_ci/utility/vsphere.py (hunk 5 above)
Comment (ebenahar) [reply ocs_ci/utility/vsphere.py:634]: same as last comment

File: tests/manage/z_cluster/nodes/test_disk_failures.py (hunk 6 above)
Comment (ebenahar) [reply tests/manage/z_cluster/nodes/test_disk_failures.py:139]: ack

File: tests/manage/z_cluster/nodes/test_disk_failures.py (hunk 7 above)
Comment (ebenahar) [reply tests/manage/z_cluster/nodes/test_disk_failures.py:183]: ack

File: ocs_ci/utility/vsphere.py [hunk 8]
```diff
@@ -468,35 +468,24 @@ def destroy_pool(self, pool, dc, cluster):
         WaitForTask(pi.Destroy())
         logger.info(f"Successfully deleted resource pool {pool}")
 
-    def remove_disk(self, vm, unit_number, datastore=True):
+    def remove_disk(self, vm, identifier, key='unit_number', datastore=True):
```
Comment (RazTamir) [inline ocs_ci/utility/vsphere.py:471]: the signature changed, please confirm no calls with old one

Comment (ebenahar) [reply ocs_ci/utility/vsphere.py:471]: correct. The only additional call was adjusted accordingly:
https://github.com/red-hat-storage/ocs-ci/pull/1245/files#diff-2c4f7dc9d7d80db5ddfb8abfbfbce05aR544

File: ocs_ci/ocs/platform_nodes.py [hunk 9]
```diff
@@ -413,7 +448,7 @@ def terminate_nodes(self, nodes, wait=True):
         )
         self.aws.terminate_ec2_instances(instances=instances, wait=wait)
 
-    def detach_volume(self, volume):
+    def detach_volume(self, volume, node=None, delete_from_backend=True):
```
Comment (MeridianExplorer) [inline ocs_ci/ocs/platform_nodes.py:451]: Need to update docstring

File: tests/manage/z_cluster/nodes/test_disk_failures.py [hunk 10]
```diff
@@ -129,3 +137,113 @@ def test_detach_attach_2_data_volumes(self, nodes, pvc_factory, pod_factory):
...
+        osd_pvcs_count = len(osd_pvcs)
+        osd_pvc = [ds for ds in osd_pvcs if ds.get().get('metadata').get('name') == claim_name][0]
+
+        # Get the corresponding OSD pod
+        logger.info(f"Getting the corresponding OSD pod of PVC {osd_pvc.name}")
+        osd_pods = get_osd_pods()
+        osd_pods_count = len(osd_pods)
+        osd_pod = [
+            osd_pod for osd_pod in osd_pods if osd_pod.get()
+            .get('metadata').get('labels').get('ceph.rook.io/pvc') == claim_name
```
Comment (MeridianExplorer) [inline tests/manage/z_cluster/nodes/test_disk_failures.py:174]: Replace `ceph.rook.io/pvc` with `CEPH_ROOK_IO_PVC_LABEL`

Comment (ebenahar) [reply tests/manage/z_cluster/nodes/test_disk_failures.py]: done

File: ocs_ci/ocs/platform_nodes.py (hunk 9 above)
Comment (ebenahar) [reply ocs_ci/ocs/platform_nodes.py:451]: done

**Comment by ebenahar** [comment]
Above PR validation job failure is on deployment validation of OSDs distribution which is done wrong for clusters that were already expanded. I have used an already existing vsphere cluster that went through add capacity test. Issue - https://github.com/red-hat-storage/ocs-ci/issues/1819. Fixed in https://github.com/red-hat-storage/ocs-ci/pull/1820.

File: tests/sanity_helpers.py [hunk 11]
```diff
@@ -24,15 +24,15 @@ def __init__(self):
...
 
-    def health_check(self, cluster_check=True):
+    def health_check(self, cluster_check=True, tries=20):
         """
         Perform Ceph and cluster health checks
         """
         logger.info("Checking cluster and Ceph health")
         node.wait_for_nodes_status()
 
         assert ceph_health_check(
```
Comment (RazTamir) [inline tests/sanity_helpers.py:34]: the assert is useless here

Comment (ebenahar) [reply tests/sanity_helpers.py]: @RazTamir thanks, removed

**Pull request merged**